File: problems/Java/124. Binary Tree Maximum Path Sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int mx = Integer.MIN_VALUE;

    public int maxPathSum(TreeNode root) {
        dfs(root);
        return mx;
    }

    private int dfs(TreeNode n) {
        if (n == null) return 0;

        int l = Math.max(0, dfs(n.left));
        int r = Math.max(0, dfs(n.right));

        mx = Math.max(mx, n.val + l + r);

        return n.val + Math.max(l, r);
    }
}

File: problems/C++/124. Binary Tree Maximum Path Sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int mx = INT_MIN;

    int maxPathSum(TreeNode* root) {
        dfs(root);
        return mx;
    }

    int dfs(TreeNode* n) {
        if (!n) return 0;

        int l = max(0, dfs(n->left));
        int r = max(0, dfs(n->right));

        mx = max(mx, n->val + l + r);

        return n->val + max(l, r);
    }
};